lib/rack/query_parser.rb

```diff
@@ -98,7 +98,11 @@ def normalize_params(params, name, v, depth)
         params[k] ||= []
         raise ParameterTypeError, "expected Array (got #{params[k].class.name}) for param `#{k}'" unless params[k].is_a?(Array)
         params[k] << v
-      elsif after =~ %r(^\[\]\[([^\[\]]+)\]$) || after =~ %r(^\[\](.+)$)
+      elsif params[k].is_a?(Array) && after =~ %r(^\[(\d+)\]$)
+        index = $1.to_i
+        raise InvalidParameterError, "skipped Array index (got #{index} while current length is #{params[k].length}) for param `#{k}'" if index > params[k].length
```

Comment:
I think you should make a new error class. It's close but it doesn't contain "invalid" parameters, rather it's missing parameters (ie you're asking for index 5 when index 5 doesn't exist).